samples/bluetooth/broadcast_assistant/src/main.c

```diff
@@ -0,0 +1,387 @@
+
+	err = k_sem_take(&sem_source_discovered, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_source_discovered (err %d)\n", err);
+	}
+}
+
+static void scan_for_broadcast_sink(void)
+{
+	int err;
+
+	scanning_for_broadcast_source = false;
+
+	err = bt_le_scan_start(BT_LE_SCAN_PASSIVE, NULL);
+	if (err) {
+		printk("Scanning failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Scanning for Broadcast Sink successfully started\n");
+
+	err = k_sem_take(&sem_sink_discovered, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_sink_discovered (err %d)\n", err);
+	}
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (err != 0) {
+		printk("Failed to connect to %s (%u)\n", addr, err);
+
+		bt_conn_unref(broadcast_sink_conn);
+		broadcast_sink_conn = NULL;
+
+		scan_for_broadcast_sink();
+		return;
+	}
+
+	if (conn != broadcast_sink_conn) {
+		return;
+	}
+
+	printk("Connected: %s\n", addr);
+	k_sem_give(&sem_sink_connected);
+}
+
+static void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	if (conn != broadcast_sink_conn) {
+		return;
+	}
+
+	(void)bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason 0x%02x)\n", addr, reason);
+
+	bt_conn_unref(broadcast_sink_conn);
+	broadcast_sink_conn = NULL;
+
+	k_sem_give(&sem_sink_disconnected);
+}
+
+static void security_changed_cb(struct bt_conn *conn, bt_security_t level,
+				enum bt_security_err err)
+{
+	if (err == 0) {
+		printk("Security level changed: %u\n", level);
+		k_sem_give(&sem_security_updated);
+	} else {
+		printk("Failed to set security level: %u\n", err);
+	}
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.connected = connected,
+	.disconnected = disconnected,
+	.security_changed = security_changed_cb
+};
+
+
+int main(void)
+{
+	int err;
+
+	err = bt_enable(NULL);
+	if (err) {
+		printk("Bluetooth init failed (err %d)\n", err);
+		return 0;
+	}
+	printk("Bluetooth initialized\n");
+
+	bt_le_scan_cb_register(&scan_callbacks);
+
+	scan_for_broadcast_sink();
+
+	err = k_sem_take(&sem_sink_connected, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_sink_connected (err %d)\n", err);
+	}
+
+	bt_bap_broadcast_assistant_discover(broadcast_sink_conn);
+
+	err = k_sem_take(&sem_security_updated, K_FOREVER);
+	if (err != 0) {
+		printk("Failed to take sem_security_updated (err %d)\n", err);
+	}
+
+	scan_for_broadcast_source();
+
+	struct bt_bap_scan_delegator_subgroup subgroup = { 0 };
+
+	subgroup.bis_sync = 0x06;
+
+	struct bt_bap_broadcast_assistant_add_src_param param = { 0 };
+
+	bt_addr_le_copy(&param.addr, &selected_addr);
+	param.adv_sid = selected_sid;
+	param.pa_interval = selected_pa_interval;
+	param.broadcast_id = selected_broadcast_id;
+	param.pa_sync = true;
+	param.num_subgroups = 1;
```

Comment:
Hardcoded value :)

Not sure what is valid to do here; e.g. if the source has 2 subgroups, can we just provide one? 

BASS states
> The Num_Subgroups field is used to expose the number of subgroups present in the Broadcast Audio Source Endpoint (BASE) structure, defined in Section 3.7.2.2 in [3], used to describe a BIG.

So I think we have to PA sync to the broadcast source to get the BASE to provide that value to the server, but the specs are actually a bit unclear on that.

Reply:
I think I will try with a few variants of sources today.  Maybe it just takes the first one found/indexed if set to 1?  I'll try with both the Samsung Galaxy Buds2 Pro and a board with this: https://github.com/zephyrproject-rtos/zephyr/pull/65014

Reply:
I don't currently have an easy way of testing with multiple subgroups. The current broadcast_audio_sample has one.  Would this be a potential future improvement.  Also, it would require doing the PA sync and more.

Reply:
Indeed. The question is whether this is valid behavior (a sample should not have known invalid behavior). 

The way I read it, we need to sync to the PA for the BASE to get the subgroup count etc. (or even if is syncable). The current iteration of this sample is built on quite a few assumptions :D 

Similarly a broadcast assistant should also verify that the sink can support the BASE before, but as discussed offline this isn't possible yet (https://github.com/zephyrproject-rtos/zephyr/issues/68283)